### PR TITLE
 D3ASIM-3348: Websocket consumer for live forecasted data.

### PR DESCRIPTION
**Pull request by faizan2590**
In order to locally test it:

- Update the environment variables under `.envvars`
- Launch the websocket server via `python live_data/websocket_subscriber/test_scripts/ws_server.py.py`
- Launch the main script to consume messages from mqtt and ws and forward the device reading to respective CNs via `python live_data/main.py`
- To test sending the mocked data to ws, update the `producer.py` script with respective device name and its energy reading and launch via `python live_data/websocket_subscriber/test_scripts/producer.py`

### CODE REVIEW

File: d3a_api_client/constants.py
Comment (spyrostz) [inline d3a_api_client/constants.py:11]: I think it is more clear named as CUSTOMER_WEBSOCKET_DOMAIN_NAME or similar

File: mqtt_subscriber/main.py
Comment (spyrostz) [inline mqtt_subscriber/main.py:18]: The problem with processes is that if one of the 2 crashes, this will not result in the mqtt_subscriber pod terminating. For instance, think of the case where the websocket process dies, and the mqtt process continues to operate. The pod will not restart, therefore the websocket process will never restart (the pod will keep on running only with the mqtt process). 
On another note, I am not sure it is going to work as expected in dev env, because the main process will terminate (there is no `join()` call for the processes as far as I see)

File: websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline websocket_subscriber/live_data_consumer.py:20]: You can skip that, and the base class constructor will be automatically invoked

File: websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline websocket_subscriber/live_data_consumer.py:35]: Since there are no commands nor events here, there is no need for 2 separate methods. A single `received_message` method should suffice, therefore you can delete the `_handle_event_message` and immediately call `self.client.set_energy_forecast(**message['data'])` from the `received_message` method.

File: websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline websocket_subscriber/live_data_consumer.py:76]: I am fairly certain that this function is already there, otherwise the mqtt subscriber would not work. Also, here you pass the responsibility of finding out the simulation_id and area_uuid to the customer, while we can find out this information (we do the same for mqtt, we match the area names from CNs that are supposed to receive measurements, and we automatically find out the simulation_id and area_uuid)

**Review by spyrostz — COMMENTED** [review]
Some comments, I'll have another look once implemented

File: integration_tests/test_aggregator_batch_commands.py
Comment (fievelk) [inline integration_tests/test_aggregator_batch_commands.py:89]: This can be safely removed :P

**Review by fievelk — APPROVED** [review]
Just a minor comment about commented code. LGTM

File: mqtt_subscriber/main.py
Comment (faizan2590) [reply mqtt_subscriber/main.py:18]: After offline discussion

- No need to have `.join` as python library do so.
- Both `MQTT`  and `websocket`  could be launch as two separate processes via main.py script.

File: live_data/main.py
Comment (spyrostz) [inline live_data/main.py:26]: I think you can make it a bit more DRY with a `create_process_nonblocking` function.

File: live_data/mqtt_subscriber/oli_broker.py
Comment (spyrostz) [inline live_data/mqtt_subscriber/oli_broker.py:51]: Please either convert these to debug logs or remove these before merging if you do not need them anymore

File: live_data/websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline live_data/websocket_subscriber/live_data_consumer.py:22]: Please rename the `area_name` to the slightly more generic `device_identifier`

File: live_data/websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline live_data/websocket_subscriber/live_data_consumer.py:24]: `energy_wh` -> energy_kWh

File: live_data/websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [inline live_data/websocket_subscriber/live_data_consumer.py:23]: Instead of constructing a new RestDeviceClient object for all messages, you can reuse the `self.client` IIRC

File: live_data/websocket_subscriber/test_scripts/ws_server.py
Comment (spyrostz) [inline live_data/websocket_subscriber/test_scripts/ws_server.py:1]: If you want to leave it for testing it's ok, however the server I am currently implementing is quite different from this implementation (mainly because there is no need to broadcast a message, and also because I have to wait 15 mins before every transmission)

File: live_data/main.py
Comment (spyrostz) [inline live_data/main.py:1]: General comment, live_data is a bit vague for the package name. A much better name IMO is `live_data_subscriber` or `live_data_consumer` to denote the fact that this service is subscribing to external data sources.

**Review by spyrostz — COMMENTED** [review]
Some comments, LGTM once fixed though

File: live_data/websocket_subscriber/__init__.py
Comment (spyrostz) [inline live_data/websocket_subscriber/__init__.py:6]: This is also not needed, a `lambda: LiveDataConsumer(device_api_client_mapping)` inside the main.py should suffice if you want to defer execution of this method because it is blocking.

File: live_data/websocket_subscriber/live_data_consumer.py
Comment (faizan2590) [reply live_data/websocket_subscriber/live_data_consumer.py:24]: `set_energy_forecast ` is accepts value in `Wh`

File: live_data/websocket_subscriber/live_data_consumer.py
Comment (spyrostz) [reply live_data/websocket_subscriber/live_data_consumer.py:23]: After offline discussion with @faizan2590 , this is not feasible because the same websocket is used for multiple devices, therefore the rest client will have to know to which device to send this measurement. Should remain as is.

File: live_data/websocket_subscriber/test_scripts/ws_server.py
Comment (faizan2590) [reply live_data/websocket_subscriber/test_scripts/ws_server.py:1]: Will leave it for future to instantly test it

File: live_data_subscriber/mqtt/oli_broker.py
Comment (faizan2590) [inline live_data_subscriber/mqtt/oli_broker.py:24]: To be removed before merging.

File: live_data_subscriber/mqtt/oli_broker.py
Comment (faizan2590) [inline live_data_subscriber/mqtt/oli_broker.py]: To be removed before merging.

**Review by spyrostz — DISMISSED** [review]
LGTM

File: live_data_subscriber/mqtt/__init__.py
Comment (fievelk) [inline live_data_subscriber/mqtt/__init__.py:30]: If the token authentication fails or there are no Canary Networks available, the script will crash at this point because of this line: https://github.com/gridsingularity/d3a-api-client/blob/7116667243cb34097d3a63855219f72b72c5f0cc/d3a_api_client/utils.py#L214

I think it would be good to add a check in `list_running_canary_networks_and_devices_with_live_data` to make sure that:
- `data` is returned (and it's not None)
- there have been no errors during the request

File: live_data_subscriber/.envvars
Comment (fievelk) [inline live_data_subscriber/.envvars:11]: Is there a specific reason why we're not using something like [python-dotenv](https://github.com/theskumar/python-dotenv) here, rather than running `set -a ; source .envvars ; set +a`?

Comment (faizan2590) [reply live_data_subscriber/.envvars:11]: Its been there before this PR, I would actually ask @spyrostz of having this as a preference.

File: live_data_subscriber/mqtt/__init__.py
Comment (fievelk) [reply live_data_subscriber/mqtt/__init__.py:30]: On this line: https://github.com/gridsingularity/d3a-api-client/blob/7116667243cb34097d3a63855219f72b72c5f0cc/live_data_subscriber/mqtt/__init__.py#L37

if the `cn_mapping` is empty (when there are no CNs) the loop won't do anything, but the method will still return `topic_api_client_mapping` and `device_api_client_mapping`. So the program will keep running but it will fail later on. Do you think this should be handled somehow, or is this expected?

Comment (faizan2590) [reply live_data_subscriber/mqtt/__init__.py:30]: > If the token authentication fails , the script will crash at this point because of this line
To me it sounds like a good thing, as authentication must correctly done to continue reading data from energy meters and to be forwarded to respective CNs.

> If there are no Canary Networks available, the script will crash at this point because of this line
It didn't crashed for me, only that the received data from respective energy meter simply isn't forwarded as the mapping is empty due to no CN

Comment (faizan2590) [reply live_data_subscriber/mqtt/__init__.py]: > the program will keep running but it will fail later on. Do you think this should be handled somehow, or is this expected?

Thanks for prolonged testing, the culprit was at this [line](https://github.com/gridsingularity/d3a-api-client/pull/113/files/6cfe04baa216f6f066e4f4d2ace87bb3dd3f4b04?file-filters%5B%5D=.py&hide-deleted-files=true#diff-2ac8c9997c6823e9284fe32ec6bfe4666ea735e13ce528bd7be42b6f7aa094d0R43), when after some some interval CN's device mapping is refreshed in order to take into account the newly added CN, earlier it was only updating the mapping for `mqtt` but after incorporating for `ws` the returned format was changed, that failed the script later, applied a fix via this [commit](https://github.com/gridsingularity/d3a-api-client/pull/113/commits/6c5cdefbd36b131b72409de614f84d7573f19987). Feel free to give another try.

Comment (fievelk) [reply live_data_subscriber/mqtt/__init__.py]: > To me it sounds like a good thing, as authentication must be correctly done to continue reading data from energy meters and to be forwarded to respective CNs.

I agree with you that the program should stop. I think it would be better if it would exit gracefully rather than crashing, though. Do you think we could show the user a message saying that something went wrong with the authentication? This would probably involve changing this part: https://github.com/gridsingularity/d3a-api-client/blob/6c5cdefbd36b131b72409de614f84d7573f19987/d3a_api_client/utils.py#L60-L64

Rather than returning `None`, we could raise an exception and then handle it where the method is called:
```python
try:
    data = execute_graphql_request(domain_name=domain_name, query=query)
except SomeCustomAuthenticationException as ex:
    # Exit gracefully with a message
    sys.exit(ex)  # The message might need to be written differently
```

What do you think?

File: d3a_api_client/utils.py
Comment (hannesdiedrich) [inline d3a_api_client/utils.py:63]: Minor, but I would not leave this in "error" level because it could potentially spam our logs.

File: live_data_subscriber/websocket/consumer.py
Comment (hannesdiedrich) [inline live_data_subscriber/websocket/consumer.py:72]: AFAIS, this is not used, right?

File: d3a_api_client/utils.py
Comment (fievelk) [reply d3a_api_client/utils.py:63]: @hannesdiedrich Isn't this logged only on the client side? Shouldn't the program fail if the authentication is not successful?

Comment (hannesdiedrich) [reply d3a_api_client/utils.py:63]: @fievelk You are right, this function is used both for the normal client and the WSComsumer which we will run on our cluster. So, leave it as is for now. If we get spammed by this error message, we can deal with it later.

File: live_data_subscriber/websocket/test_scripts/ws_server.py
Comment (hannesdiedrich) [inline live_data_subscriber/websocket/test_scripts/ws_server.py:9]: Please find a less generic way for this class. 
I would say at least `WSServer`.

File: live_data_subscriber/websocket/test_scripts/ws_server.py
Comment (hannesdiedrich) [inline live_data_subscriber/websocket/test_scripts/ws_server.py:42]: Aren't these two different handlers? Or can both functionalities be combined?
Copy/Paste from the asyncio.events:
```
    def run_forever(self):
        """Run the event loop until stop() is called."""
        raise NotImplementedError

    def run_until_complete(self, future):
        """Run the event loop until a Future is done.

        Return the Future's result, or raise its exception.
        """
        raise NotImplementedError
```

File: live_data_subscriber/websocket/test_scripts/ws_server.py
Comment (hannesdiedrich) [inline live_data_subscriber/websocket/test_scripts/ws_server.py:46]: I dont find a place in the codediff where you actually use this whole module. Can you help me please?

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [inline live_data_subscriber/__init__.py:83]: You have `api_client_dict` as input variable, but you never alter it. Is it really needed?

Comment (hannesdiedrich) [reply live_data_subscriber/__init__.py:83]: Also, I dont see the need for returning `last_time_checked`. Were is it used outside of this function?

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [inline live_data_subscriber/__init__.py:35]: so for the future, a new `allowed_devices_name_mapping` has to be created, rigth? For that we need information from the costumer, right?

Comment (hannesdiedrich) [reply live_data_subscriber/__init__.py:35]: How about making this an input variable and only returning one `device_api_client_mapping` (not a tuple).
This way you can decide in `refresh_cn_and_device_list` which mapping is used.

File: live_data_subscriber/websocket/test_scripts/ws_server.py
Comment (hannesdiedrich) [reply live_data_subscriber/websocket/test_scripts/ws_server.py:9]: OK, thanks for the offline discussion. This is minor then (as this is only for testing)

**Review by hannesdiedrich — COMMENTED** [review]
Left some comments, happy to review again.

File: live_data_subscriber/__init__.py
Comment (faizan2590) [reply live_data_subscriber/__init__.py:35]: > so for the future, a new allowed_devices_name_mapping has to be created, rigth? For that we need information from the costumer, right?

May be, not sure, probably depends on client's requirement to either support both mqtt as well as ws or one of it for their devices, for later case having a separate `allowed_devices_name_mapping ` for mqtt and ws sounds better.

> How about making this an input variable...
Sorry didn't completely get you ,Input variable to what?

File: live_data_subscriber/websocket/test_scripts/ws_server.py
Comment (faizan2590) [reply live_data_subscriber/websocket/test_scripts/ws_server.py:42]: After testing it both were needed. Since Im implementing it for the first time and not yet an expert so may not have the best answer for your question, `run_until_complete `  is an asynchronous iterator that do the job for objects being passed whereas `run_forever ` is to hold the script from exiting, so you would see `run_forever ` in `server` and `consumer´ script but not in `producer` script.

File: live_data_subscriber/mqtt/__init__.py
Comment (faizan2590) [reply live_data_subscriber/mqtt/__init__.py]: Im not sure if gracefully shutting would help k8s to restart the service on its own. @spyrostz @hannesdiedrich ?

File: live_data_subscriber/websocket/consumer.py
Comment (faizan2590) [reply live_data_subscriber/websocket/consumer.py:72]: yes, you  are right, removing it

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [reply live_data_subscriber/__init__.py:35]: What i meant is that you are performing the same steps for creating `device_api_client_mapping` and `topic_api_client_mapping`. The only difference in the future will be the `allowed_devices_name_mapping`, so why not making this an input variable for `generate_api_client_args_mapping`?

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [reply live_data_subscriber/__init__.py:83]: I see, you pulled this out of the `MQTTConnection` class to use it also in `LiveDataWebsocketMessageReceiver`. 
IMO how it is implemented now it is not really pretty.I would vote for change by reference and omit the return values here. But this violates the guideline we agreed on (prefer return values over change by reference) and may also be premature. 
So, leave as is then

File: live_data_subscriber/__init__.py
Comment (faizan2590) [reply live_data_subscriber/__init__.py]: Adapted via https://github.com/gridsingularity/d3a-api-client/pull/113/commits/b80f0a53a24d3d51f97f66d5b6d470c9897248f9

File: live_data_subscriber/mqtt/__init__.py
Comment (spyrostz) [reply live_data_subscriber/mqtt/__init__.py]: @faizan2590 no matter how the service is terminated the service will be restarted by K8s, but I see little help in that regard, since IIRC sys.exit also raises an exception `SystemExit` , therefore the application will still terminate with an unhandled exception, and I suspect that the annoying traceback will still be present. IMO the only advantage in terminating gracefully in this case is to mute the traceback from an exception that you do not want to see. 

Give it a try, but if the traceback is still reported in this case I would personally prefer to not use sys.exit at all, because there is no gain over what we have now.

Comment (fievelk) [reply live_data_subscriber/mqtt/__init__.py]: I think I might be missing some part of the broader view for this PR: since I am running the script locally as described in the description of the PR, I'm not aware of the role of k8s. Exiting gracefully would only make sense if this script was used by a user that might need some kind of and readable output. If it's part of an automated pipeline, this is not so important I guess :)

File: live_data_subscriber/.envvars
Comment (spyrostz) [reply live_data_subscriber/.envvars:11]: Because this is supposed to be a convenient function for local testing, not a way to set configuration in real operation of the live_data subscriber. These env variables are all supposed to be handled by the K8 config map in the deployed live_data_subscriber, therefore the .env file would not be useful. Therefore I would be very hesitant to add a dependency + Python code that reads .env files without any need for such code in production, especially since it is an one-liner to read the file locally.

File: live_data_subscriber/mqtt/__init__.py
Comment (spyrostz) [reply live_data_subscriber/mqtt/__init__.py]: This is essentially the subscriber for the data reported for Simris. Running the script locally is mainly for testing purposes, and not by users other than us. The main purpose of this script is to live in our cluster, connect to the Simris server via websocket to read the measurements for their devices, and propagate these measurements to d3a.

Comment (faizan2590) [reply live_data_subscriber/mqtt/__init__.py]: `retrieve_jwt_key_from_server ` is being used in many other places, if I change it now from returning `None` to raising an exception other parts of codes has to be adapted too. I would propose to create an small ticket for that as it seems to deviates from the main purpose of this ticket.

**Comment by hannesdiedrich** [comment]
@faizan2590 I meant:
Lets have only the `allowed_devices_name_mapping` decide how the `key_name` is populated in `generate_api_client_args_mapping`. For now is suggest to have two different `allowed_devices_name_mapping`, onw for the MQTT and one for the WS.
for the WS it could be a temporary `{dev_name1: dev_name1, dev_name2: dev_name2, ..}` until we know better from the customer.

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [inline live_data_subscriber/__init__.py:66]: you dont need the mqtt flag any more! That was the point actually

Comment (faizan2590) [reply live_data_subscriber/__init__.py:66]: Please give this commit another look: https://github.com/gridsingularity/d3a-api-client/commit/381250b20a941f95ba8c06591e3b13b6fd21d4a5
`is_mqtt`  is no more used inside `generate_api_client_args_mapping`

File: live_data_subscriber/__init__.py
Comment (hannesdiedrich) [inline live_data_subscriber/__init__.py:60]: Isnt it rather a `allowed_device_mapping` than a `default_api_client_map` ? 
```suggestion
                               allowed_device_mapping):
```

**Review by hannesdiedrich — DISMISSED** [review]
One last comment, once addressed, LGTM

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Review by spyrostz — APPROVED** [review]
LGTM

**Pull request merged**
